### PR TITLE
45802492 - Simple Diffusion

### PR DESCRIPTION
**Pull request by michaelholmes4**
Name: Michael Holmes
SN: 45802492

Simple diffusion image generation library created in PyTorch for COMP3710 project assessment.

Contents:
* `train.py` - Command line utility that trains a new diffusion model on a dataset.
* `dataset.py` - Wraps a directory of image files in a PyTorch dataloader. Images can be any size or format that can be opened by PIL. All images are resized to a given dimension, converted to RGB and normalised to a range of -1 to 1.
* `modules.py` - Contains a Trainer class to handle training of the model. Contains the U-Net model and required components.
* `predict.py` - Command line utility to predict a new images from an existing `.pth` model

Includes an `environment.yml` to setup conda dependencies.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**
- Readme: 
- Results: OK,
- Code: OK,
- Commit messages: OK but some should be more specific

IMPORTANT: Commits are not in the right place. Did you create the topic-recognition branch yourself? You should fork Shakes repo in whole (there is a checkmark when you fork) and commit to his topic-recognition branch. You should be able to see work from the previous years in "recognition folder"

**Comment by michaelholmes4** [comment]
Yes I created topic-recognition myself, thats how I read the assignment sheet since you can create new branch using checkout command.
Screen Shot 2022-11-09 at 9 09 32 pm

 I did not know it was possible to fork entire repo. 

As for commits being in the wrong place I did my work using the template files in the recognition folder before realising they had to be moved into a seperate student folder. See the post I made on ed:
https://edstem.org/au/courses/9608/discussion/1083590
As such there is no way to fix that, but I was told that would be OK.

**Comment by SiyuLiu0329** [comment]
@michaelholmes4 
It is okay, git is extremely flexible and you can indeed migrate all the commits to the correct branch. That is why we have marks for fixing errors in PRs.
One possible solution would be something like 
1.  put your current work in a separate branch e.g. "new_branch"
2. try pulling the original pattern-recognition branch, delete the one you have already made if necessary, as long as you have a backup of all the code
3. git checkout  the original pattern-recognition branch
4. git merge "new_branch"

**Comment by michaelholmes4** [comment]
Ok thanks. I have just merged topic-recognition into my branch so it is now up to date with the upstream branch and all previous years work is there. Will that be ok?

**Comment by michaelholmes4** [comment]
Ok I have done two things. I have merged pattern-recognition from upstream into my own branch so that they are now in sync and are effectively the same branch. All submissions from previous years are visible. 

 I have also restored some files that were modified as a result of using my own branch.

Please let me know if there is anything else.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages -1 (could be more specific)
Progressive commits used
## Documentation
ReadMe acceptable/good
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
